modules/mcc/src/utils.cpp

```diff
@@ -0,0 +1,120 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+//
+//
+//                       License Agreement
+//              For Open Source Computer Vision Library
+//
+// Copyright(C) 2020, Huawei Technologies Co.,Ltd. All rights reserved.
+// Third party copyrights are property of their respective owners.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//             http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Author: Longbu Wang <wanglongbu@huawei.com.com>
+//         Jinheng Zhang <zhangjinheng1@huawei.com>
+//         Chenqi Shan <shanchenqi@huawei.com>
+
+#include "utils.hpp"
+
+namespace cv
+{
+namespace ccm
+{
+
+double gammaCorrection_(const double& element, const double& gamma)
+{
+    return (element >= 0 ? pow(element, gamma) : -pow((-element), gamma));
+}
+
+cv::Mat gammaCorrection(const cv::Mat& src, const double& gamma)
+{
+    return elementWise(src, [gamma](double element)->double {return gammaCorrection_(element, gamma); });
+}
+
+cv::Mat maskCopyTo(const cv::Mat& src, const cv::Mat& mask)
+{
+    cv::Mat dst(countNonZero(mask), 1, src.type());
+    const int channel = src.channels();
+    auto it_mask = mask.begin<uchar>();
+    switch (channel)
+    {
+    case 1:
+    {
+        auto it_src = src.begin<double>(), end_src = src.end<double>();
+        auto it_dst = dst.begin<double>();
+        for (; it_src != end_src; ++it_src, ++it_mask)
+        {
+            if (*it_mask)
+            {
+                (*it_dst) = (*it_src);
+                ++it_dst;
+            }
+        }
+        break;
+    }
+    case 3:
+    {
+        auto it_src = src.begin<cv::Vec3d>(), end_src = src.end<cv::Vec3d>();
+        auto it_dst = dst.begin<cv::Vec3d>();
+        for (; it_src != end_src; ++it_src, ++it_mask)
+        {
+            if (*it_mask)
+            {
+                (*it_dst) = (*it_src);
+                ++it_dst;
+            }
+        }
+        break;
+    }
+    default:
+        throw std::invalid_argument { "Wrong channel!" };
+        break;
+    }
+    return dst;
+}
+
+cv::Mat multiple(const cv::Mat& xyz, const cv::Mat& ccm)
+{
+    cv::Mat tmp = xyz.reshape(1, xyz.rows * xyz.cols);
+    cv::Mat res = tmp * ccm;
+    res = res.reshape(res.cols, xyz.rows);
```

Comment:
`operator *` invokes `cv::gemm()` by default.

Instead of playing with `.reshape` it is better to call `cv::multiply(xyz, ccm, res)` explicitly.

Reply:
Thanks, but we should make matrix multiplication between xyz and ccm, not multiply calculates the per-element product, calling `cv::multiply` maybe unsuitable.  And, before making matrix multiplication, the xyz matrix should have the same col with ccm's row, so `.reshape`  is needed.